streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorMetadata.java

```diff
@@ -0,0 +1,153 @@
+    }
+
+    public static ProcessorMetadata deserialize(final byte[] metaData) {
+        if (metaData == null || metaData.length == 0) {
+            return new ProcessorMetadata();
+        }
+
+        final ByteBuffer buffer = ByteBuffer.wrap(metaData);
+        final int entrySize = buffer.getInt();
+        final Map<String, Long> metadata = new HashMap<>(entrySize);
+        for (int i = 0; i < entrySize; i++) {
+            final int keySize = buffer.getInt();
+            final byte[] keyBytes = new byte[keySize];
+            buffer.get(keyBytes);
+            final Long value = buffer.getLong();
+            metadata.put(new String(keyBytes, StandardCharsets.UTF_8), value);
+        }
+        return new ProcessorMetadata(metadata);
+    }
+
+    public byte[] serialize() {
+        if (metadata.isEmpty()) {
+            return new byte[0];
+        }
+
+        int kvSize = 0;
+        for (final Map.Entry<String, Long> entry : metadata.entrySet()) {
+            kvSize += Integer.BYTES;
+            kvSize += entry.getKey().getBytes(StandardCharsets.UTF_8).length;
+            kvSize += Long.BYTES;
+        }
+
+        final int capacity = Integer.BYTES + kvSize;
+        final ByteBuffer buffer = ByteBuffer.allocate(capacity).putInt(metadata.size());
+        for (final Map.Entry<String, Long> entry : metadata.entrySet()) {
+            final byte[] keyBytes = entry.getKey().getBytes(StandardCharsets.UTF_8);
+            final int keyLen = keyBytes.length;
+            buffer.putInt(keyLen)
+                .put(keyBytes)
+                .putLong(entry.getValue());
+        }
+        return buffer.array();
+    }
+
+    public void addMetadata(final String key, final long value) {
+        metadata.put(key, value);
+        needsCommit = true;
+    }
+
+    public Long getMetadata(final String key) {
+        return metadata.get(key);
+    }
+
+    /**
+     * Merge with other metadata. Missing keys will be added. Existing key's value will be updated to
+     * max
+     * @param other Other metadata to be merged
+     */
+    public void update(final ProcessorMetadata other) {
+        if (other == null) {
+            return;
+        }
+        for (final Map.Entry<String, Long> kv : other.metadata.entrySet()) {
+            final Long value = metadata.get(kv.getKey());
+            if (value == null || value < kv.getValue()) {
+                metadata.put(kv.getKey(), kv.getValue());
+            }
+        }
+    }
+
+    public void setNeedsCommit(final boolean needsCommit) {
+        this.needsCommit = needsCommit;
+    }
+
+    /**
+     * Whether metadata needs to be committed. It should be committed only if addMetadata is or
+     * setNeedsCommit is called explicitly
+     *
+     * @return If metadata needs to be committed.
+     */
+    public boolean needCommit() {
+        return needsCommit;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(metadata);
```

Comment:
I assume it's on purpose to omit `needsCommit` flag? Might be worth to add a comment.